pkg/cloudprovider/provider/azure/provider.go

```diff
@@ -0,0 +1,615 @@
+	// orphaned NICs, IPs and disks in case the operation is interrupted in the middle.
+	glog.Infof("deleting VM %q", machine.Name)
+	if err = deleteVMsByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove public IP addresses of machine %q: %s", machine.Name, err)
+	}
+
+	glog.Infof("deleting disks of VM %q", machine.Name)
+	if err = deleteDisksByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove disks of machine %q: %s", machine.Name, err)
+	}
+
+	glog.Infof("deleting network interfaces of VM %q", machine.Name)
+	if err = deleteInterfacesByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove network interfaces of machine %q: %s", machine.Name, err)
+	}
+
+	glog.Infof("deleting public IP addresses of VM %q", machine.Name)
+	if err = deleteIPAddressesByMachineUID(context.TODO(), config, machine.UID); err != nil {
+		return fmt.Errorf("failed to remove public IP addresses of machine %q: %s", machine.Name, err)
+	}
+
+	return nil
+}
+
+func getVMByUID(ctx context.Context, c *config, uid types.UID) (*compute.VirtualMachine, error) {
+	vmClient, err := getVMClient(c)
+	if err != nil {
+		return nil, err
+	}
+
+	list, err := vmClient.ListAll(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var allServers []compute.VirtualMachine
+
+	for {
+		allServers = append(allServers, list.Values()...)
+		if list.NotDone() {
+			list.Next()
+		} else {
+			break
+		}
+	}
+
+	for _, vm := range allServers {
+		if vm.Tags != nil && *vm.Tags[machineUIDTag] == string(uid) {
+			return &vm, nil
+		}
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
+}
+
+func getVMStatus(ctx context.Context, c *config, vmName string) (instance.Status, error) {
+	vmClient, err := getVMClient(c)
+	if err != nil {
+		return instance.StatusUnknown, err
+	}
+
+	iv, err := vmClient.InstanceView(ctx, c.ResourceGroup, vmName)
+	if err != nil {
+		return instance.StatusUnknown, fmt.Errorf("failed to get instance view for machine %q: %s", vmName, err)
+	}
+
+	if iv.Statuses == nil {
+		return instance.StatusUnknown, nil
+	}
+
+	// it seems that this field should contain two entries: a provisioning status and a power status
+	if len(*iv.Statuses) < 2 {
+		provisioningStatus := (*iv.Statuses)[0]
+		if provisioningStatus.Code == nil {
+			glog.Warningf("azure provisioning status has missing code")
+			return instance.StatusUnknown, nil
+		}
+
+		switch *provisioningStatus.Code {
+		case "":
+			return instance.StatusUnknown, nil
+		case "ProvisioningState/deleting":
+			return instance.StatusDeleting, nil
+		default:
+			glog.Warningf("unknown Azure provisioning status %q", *provisioningStatus.Code)
+			return instance.StatusUnknown, nil
+		}
+	}
+
+	// the second field is supposed to be the power status
+	// https://docs.microsoft.com/en-us/azure/virtual-machines/windows/tutorial-manage-vm#vm-power-states
+	powerStatus := (*iv.Statuses)[1]
```

Comment:
the length of the slice should previously checked ;)

Reply:
Because previous branch will return, when len < 2

Reply:
Oh, you're right